### PR TITLE
Error Handling: Add NotAMachOError for edge cases.

### PR DESCRIPTION
**Pull request by woodruffw**
There are two edge cases early-on in the parsing process:
files too small to have a valid magic number and files
with valid magic but invalid contents (most notably, Java
classfiles).

The NotAMachOError class covers both of these cases, and is raised
in FatFile#get_fat_header and MachOFile#get_and_check_magic.

cc @UniqMartin

### CODE REVIEW

File: lib/macho/macho_file.rb [hunk 1]
```diff
@@ -329,10 +329,13 @@ def get_mach_header
 
     # Read just the file's magic number and check its validity.
     # @return [Fixnum] the magic
+    # @raise [MachO::NotAMachOError] if the file is too small to have valid magic
     # @raise [MachO::MagicError] if the magic is not valid Mach-O magic
     # @raise [MachO::FatBinaryError] if the magic is for a Fat file
     # @private
     def get_and_check_magic
+      raise NotAMachOError.new("file is too small to be a Mach-O") if @raw_data.size < 4
```
Comment (UniqMartin) [inline lib/macho/macho_file.rb]: I think the limit should be 28 bytes (the size of `MachHeader`) instead of 4. I'd like to see a similar check in `FatFile#get_fat_header` with a limit of 8 bytes (the size of `FatHeader`; or 8+20+28 bytes if we consider a fat binary with no embedded Mach-O binaries to be invalid). See also my (yet to be written) non-code comment for more details.

File: test/test_macho.rb [hunk 2]
```diff
@@ -5,6 +5,12 @@
 class MachOFileTest < Minitest::Test
   include Helpers
 
+  def test_empty_file
+    assert_raises MachO::NotAMachOError do
+      MachO::MachOFile.new_from_bin("\x00")
```
Comment (UniqMartin) [inline test/test_macho.rb]: I think this should be just `""` (an empty string) to stay true to the name of the test.

File: lib/macho/macho_file.rb (hunk 1 above)
Comment (woodruffw) [inline lib/macho/macho_file.rb]: This makes sense.

As a side note, do you know if `clang`/any part of the OS X toolchain actually ever produces just the Mach-O header (with `ncmds` and `sizeofcmds` = 0) and nothing else?

File: test/test_macho.rb (hunk 2 above)
Comment (woodruffw) [inline test/test_macho.rb]: Good point, changing it now :+1:

**Comment by UniqMartin** [comment]
**Exception Classes.** Can we make it such that there is a single exception class that can be caught for cases where the file was successfully read, but turned out not to be a Mach-O or fat binary? I can think of three possible hierarchies that make sense to me (ordered by personal preference):

```
-+ NotAMachOError        # common ancestor, but not actually raised
 |-- MagicError          # used as before
 |-- TruncatedFileError  # too short
 |-- JavaClassFileError  # Java class file
 \-- …                   # more exceptions for cases we haven't thought of yet

-+ NotAMachOError        # too short, Java class file, …
 \-- MagicError          # used as before

-+ NotAMachOError        # common ancestor, but not actually raised
 |-- MagicError          # used as before
 \-- NonMagicError       # too short, Java class file, …
```

The important thing is that it would be sufficient to rescue `NotAMachOError` and interpret it as “regular file” (the use case in Homebrew's extension of `Pathname`), no matter what the exact cause is and without ignoring actual issues (e.g. due to unknown or misinterpreted load commands).

----

**Additional Corner Cases.** Can you add the following test cases and adjust the library code such that they succeed? (You might need to adjust the names of the exception classes per the first part of this comment.) In `test/test_fat.rb`:

```ruby
def test_empty_file
  assert_raises MachO::NotAMachOError do
    MachO::FatFile.new_from_bin("")
  end
end

def test_truncated_file
  assert_raises MachO::NotAMachOError do
    MachO::FatFile.new_from_bin("\xCA\xFE\xBA\xBE\x00\x00")
  end
end
```

In `test/test_macho.rb`:

```ruby
def test_truncated_file
  assert_raises MachO::NotAMachOError do
    MachO::MachOFile.new_from_bin("\xFE\xED\xFA\xCE\x00\x00")
  end
end
```

File: test/test_macho.rb (hunk 2 above)
Comment (UniqMartin) [inline test/test_macho.rb]: :cool:

**Comment by woodruffw** [comment]
The first hierarchy looks good to me!

File: lib/macho/macho_file.rb (hunk 1 above)
Comment (UniqMartin) [inline lib/macho/macho_file.rb]: I'm not aware of a way to create such a file except of hand-crafting it.

**Comment by woodruffw** [comment]
> MachO::FatFile.new_from_bin("\xCA\xFE\xBA\xBE\x00\x00")

`FatFile` doesn't actually have a `new_from_bin` method, so I'm going to add another truncated classfile to *test/bin* (and add the missing method in a later commit).

File: lib/macho/exceptions.rb [hunk 3]
```diff
@@ -3,14 +3,36 @@ module MachO
...
+    # @param error [String] the error in question
+    def initialize(error)
+      super error
+    end
+  end
+
+  # Raised when a file is too short to be a valid Mach-O file.
+  class TruncatedFileError < NotAMachOError
+    def initialize
+      "File is too short to be a valid Mach-O"
```
Comment (UniqMartin) [inline lib/macho/exceptions.rb]: The `super` is missing here.

File: lib/macho/exceptions.rb [hunk 4]
```diff
@@ -3,14 +3,36 @@ module MachO
...
     # @param num [Fixnum] the unknown number
     def initialize(num)
       super "Unrecognized Mach-O magic: 0x#{"%02x" % num}"
     end
   end
 
+  # Raised when a file is a Java classfile instead of a fat Mach-O.
+  class JavaClassFileError < NotAMachOError
+    def initialize
+      "File is a Java class file"
```
Comment (UniqMartin) [inline lib/macho/exceptions.rb]: The `super` is missing here, too.

File: lib/macho/exceptions.rb (hunk 3 above)
Comment (woodruffw) [inline lib/macho/exceptions.rb]: Fixed, both this and below.

File: test/test_fat.rb [hunk 5]
```diff
@@ -5,6 +5,40 @@
 class FatFileTest < Minitest::Test
   include Helpers
 
+  def test_empty_file
+    empty_file = tempfile_with_data("empty_file", "")
+
+    assert_raises MachO::NotAMachOError do
```
Comment (UniqMartin) [inline test/test_fat.rb]: Maybe check for the more specific exception classes like `TruncatedFileError` here and in the other newly added test cases?

File: test/test_fat.rb [hunk 6]
```diff
@@ -5,6 +5,40 @@
...
+    truncated_file.unlink
+  end
+
+  def test_java_classfile
+    fake_java_class_file = tempfile_with_data("fake_java_class_file", "\xCA\xFE\xBA\xBE\x00\x00\x00\x33therestofthisfileisnotarealjavaclassfile")
+
+    assert_raises MachO::NotAMachOError do
+      MachO::FatFile.new(fake_java_class_file)
+    end
+
```
Comment (UniqMartin) [inline test/test_fat.rb]: This empty line is odd and not in line with the other methods.

File: test/test_macho.rb [hunk 7]
```diff
@@ -5,6 +5,29 @@
...
+    empty_file.unlink
+  end
+
+  def test_truncated_file
+    truncated_file = tempfile_with_data("truncated_file", "\xFE\xED\xFA\xCE\x00\x00")
+
+    assert_raises MachO::NotAMachOError do
+      MachO::MachOFile.new(truncated_file)
+    end
+
```
Comment (UniqMartin) [inline test/test_macho.rb]: This empty line is odd and not in line with the other methods.

File: test/test_fat.rb [hunk 8]
```diff
@@ -5,6 +5,40 @@
...
 
+  def test_empty_file
+    empty_file = tempfile_with_data("empty_file", "")
+
+    assert_raises MachO::NotAMachOError do
+      MachO::FatFile.new(empty_file)
+    end
+  ensure
+    empty_file.close
+    empty_file.unlink
```
Comment (UniqMartin) [inline test/test_fat.rb]: I think it would be good to get rid of this code duplication in all the `ensure`s. An untested idea would be to modify `tempfile_with_data` to remember the generated file (in an instance variable) and then clean it up in the `teardown` method (not sure how to put that into `Helpers` correctly so that test classes can still add their own `teardown` methods).

**Comment by UniqMartin** [comment]
> `FatFile` doesn't actually have a `new_from_bin` method, so I'm going to add another truncated classfile to *test/bin* (and add the missing method in a later commit).

Totally missed that `FatFile` doesn't have that method, sorry. I think adding it (in a later commit) is still a good idea as it will make the interface for both classes more similar.

> Ended up creating a `Helpers.tempfile_with_data` to handle these tests.

That's also a good solution and probably more broadly applicable.

File: test/test_fat.rb (hunk 8 above)
Comment (woodruffw) [inline test/test_fat.rb]: I agree. I could probably turn `tempfile_with_data` into a block, that would reduce the repetition and avoid confusion with the interaction between `Helpers` and each test class.

Comment (UniqMartin) [inline test/test_fat.rb]: Yes, turning `tempfile_with_data` into a method that accepts a block that performs the actual test that involves the temporary file also sounds like a good solution.

**Comment by woodruffw** [comment]
The new `Helpers.tempfile_with_data` takes a block that handles setup and teardown for each temp file, so that should less of a hassle to work with.

File: test/test_fat.rb [hunk 9]
```diff
@@ -5,6 +5,30 @@
...
+  def test_truncated_file
+    tempfile_with_data("truncated_file", "\xCA\xFE\xBA\xBE\x00\x00") do |truncated_file|
+      assert_raises MachO::NotAMachOError do
+        MachO::FatFile.new(truncated_file)
+      end
+    end
+  end
+
+  def test_java_classfile
+    tempfile_with_data("fake_java_class_file", "\xCA\xFE\xBA\xBE\x00\x00\x00\x33therestofthisfileisnotarealjavaclassfile") do |fake_java_class_file|
```
Comment (UniqMartin) [inline test/test_fat.rb]: Maybe split this into two lines to avoid an awfully long line?

```ruby
blob = "\xCA\xFE\xBA\xBE\x00\x00\x00\x33therestofthisfileisnotarealjavaclassfile"
tempfile_with_data("fake_java_class_file", blob) do |fake_java_class_file|
```

Comment (woodruffw) [inline test/test_fat.rb]: Yeah, fixing it now.

File: test/helpers.rb [hunk 10]
```diff
@@ -44,4 +45,17 @@ def filechecks(except = nil)
...
+    file.write(data)
+    file.rewind
+
+    begin
+      yield file
+    ensure
+      file.close
+      file.unlink
+    end
+  end
```
Comment (UniqMartin) [inline test/helpers.rb]: I think you can simplify this further to (the only difference being that the file won't be immediately deleted, see the `Tempfile` documentation):

```ruby
def tempfile_with_data(filename, data)
  Tempfile.open(filename) do |file|
    file.write(data)
    file.rewind
    yield file
  end
end
```

Comment (woodruffw) [inline test/helpers.rb]: That's much cleaner, thanks!

**Comment by UniqMartin** [comment]
There is one (now hidden) code comment about the exception classes in the test suite, that you haven't addressed or replied to, but that's not at all a blocker, just a detail:

> Maybe check for the more specific exception classes like `TruncatedFileError` (instead of their common ancestor) here and in the other newly added test cases?

Overall I'm quite happy with how things look by now. Feel free to merge if you and the CI are happy. :+1:

**Comment by woodruffw** [comment]
> Maybe check for the more specific exception classes like TruncatedFileError (instead of their common ancestor) here and in the other newly added test cases?

Oops, missed that. That's an important detail.

After that (and the CI passes), I'll merge :smile:

**Comment by UniqMartin** [comment]
> That's an important detail.

I'm glad that we agree on this. :smile_cat:

> After that (and the CI passes), I'll merge :smile:

:+1: :clap:

**Comment by woodruffw** [comment]
(Afternote: I added `new_from_bin` in 9cdca3f7a012e0b7fded8b018bc3c9c073a7a250)
